lib/pmdtester/builders/rule_set_builder.rb

```diff
@@ -158,5 +118,82 @@ def write_dynamic_file(rule_refs)
       @options.base_config = PATH_TO_DYNAMIC_CONFIG
       @options.patch_config = PATH_TO_DYNAMIC_CONFIG
     end
+
+    private
+
+    def determine_categories_rules(filenames)
+      regression_test_required = false
+      categories = Set[]
+      rules = Set[]
+      filenames.each do |filename|
+        matched = check_single_filename(filename, categories, rules)
+        regression_test_required = true if matched
+
+        next if matched
+
+        logger.info "Change in file #{filename} doesn't match specific rule/category - enable all rules"
+        regression_test_required = true
+        categories.clear
+        rules.clear
+        break
+      end
+      [regression_test_required, categories, rules]
+    end
+
+    def check_single_filename(filename, categories, rules)
+      logger.debug "Checking #{filename}"
+
+      # matches Java-based rule implementations
+      match_data = %r{.+/src/main/java/.+/lang/([^/]+)/rule/([^/]+)/([^/]+)Rule.java}.match(filename)
```

Comment:
This is the place, where we try to map from the changed files in the PR to the changed rules. But this regex works even if the projects are a bit restructured, as proposed in pmd/pmd#3846 . As long as the structure inside a language stays...